src/components/checkbox/GroupableCheckboxReducers.ts

```diff
@@ -0,0 +1,103 @@
+import * as _ from 'underscore';
+import {IReduxAction} from '../../utils/ReduxUtils';
+import {ICheckboxState} from './CheckboxReducers';
+import {GroupableCheckboxActions, IGroupableCheckboxActionPayload} from './GroupableCheckboxActions';
+import {addChildCheckbox, addParentCheckbox, removeChildCheckbox, removeParentCheckbox} from './GroupableCheckboxReducersUtils';
+
+export interface IGroupableCheckboxesState {
+  total: number;
+  nbChecked: number;
+  parentId?: string;
+  parent?: ICheckboxState;
+  checkboxes: ICheckboxState[];
+}
+
+export const groupableCheckboxInitialState: IGroupableCheckboxesState = {total: 0, nbChecked: 0, checkboxes: []};
+export const groupableCheckboxesInitialState: IGroupableCheckboxesState[] = [];
+
+export const groupableCheckboxReducer = (state: IGroupableCheckboxesState = groupableCheckboxInitialState,
+                                         action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      if (action.payload.isParent) {
+        return addParentCheckbox(state, action);
+      }
+      return addChildCheckbox(state, action);
+
+    case GroupableCheckboxActions.toggleGroup:
+      if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+        let isChecked: boolean;
+        if (action.payload.isParent) {
+          isChecked = !state.parent.checked;
+          return {
+            ...state,
+            parent: {
+              id: action.payload.id,
+              checked: isChecked,
+            },
+            checkboxes: _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+              checkbox.checked = isChecked;
+              return checkbox;
+            }),
+            nbChecked: isChecked ? state.total : 0,
+          };
+        }
+        const checkboxes: ICheckboxState[] = _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+          if (checkbox.id === action.payload.id) {
+            isChecked = !checkbox.checked;
+            return {
+              id: checkbox.id,
+              checked: isChecked,
+            };
+          }
+          return checkbox;
+        });
+        if (checkboxes) {
+          const nbChecked: number = state.nbChecked += (isChecked ? 1 : -1);
+          return {
+            ...state,
+            parent: {
+              id: action.payload.parentId,
+              checked: nbChecked === state.total,
+            },
+            nbChecked,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
```

Comment:
We use switch/case in reducers because it is easy to see what is being returned. Here, things are getting a little blurry. Maybe you can simplify but moving the inside of your if statements in functions or something similar.